### DISCUSSION TITLE
Umi & Bigfish 想法征集 10

### DISCUSSION THREAD

**Discussion by sorrycc** [umijs/umi#11045] [Ideas]
> Why this issue? 计划在 Umi Contributor 群增加一条规则。每人每周贡献 2 个改进 Umi 或 Bigfish 的想法（辟起、辟殊、云谦），外包同学减半（1 个），外部 Contributor 随意，我每周三会在群里发个 discussion 的链接做收集。

直接回复即可。（含内部敏感信息的可以私发我）

**Comment by sorrycc** [comment]
1、[GitHub - Nozbe/eslint-plugin-nozbe: Helpful JavaScript, React, and React Native rules for ESLint, by Nozbe](https://github.com/nozbe/eslint-plugin-nozbe) 这里的一些规则不错，比如 max line、有期限的 todo、no in express、no && 等
2、考虑 Icons 和 styled-components 结合使用的场景，现在 include 的方式配置的，比较挫

**Comment by PeachScript** [comment]
1. qiankun 插件是否要把 styled-components 的 namespace 开起来避免微应用之间冲突：https://styled-components.com/docs/tooling#namespace cc @kuitos 
2. Stylelint 规则集成 styled-components 的规则：https://styled-components.com/docs/tooling#stylelint

**Comment by hanzebang** [comment]
💡 Error Code
- Error Code 结构
  - 拟定使用 8 位错误码，比如：01010001
    - 前2位：代码归属大类，比如：
      - 00：umi 核心代码
      - 01：umi 插件代码
      - 02：plugins/max 最佳实践相关代码
      - 10：bigfish 框架代码
      - ……
    - 中2位：代码分类，比如以插件维度划分：
      - 0201：access 插件
      - 0202：analytics 插件
      - 0203：antd 插件
      - ……
    - 后4位：错误序号，或各功能模块自行安排（比如00xx-警告类，10xx-错误类）
      - 02030001：Can't find antd package. Please install antd first.
      - 02030002：Can't set antd.import while using antd5 (xxx)
      - ……
- 后续消费
  - Bigfish 内部可以为以上报错补充详细的中文文档，统一提示等
- One more thing
  - 通用的用户代码修改功能，之后解法唯一的问题可以自动帮用户修改掉等

**Reply by fz6m** [reply]
这个地方是不是有些长了，前面大类感觉可以用字母区分，比如 `E0001` / `A0002` 等。

**Comment by CJY0208** [comment]
umi 中的自定义 process.env 环境变量注入目前似乎只能通过 `defineConfig({ define: {'process.env': { ... }} })` 进行

参考其他框架，可以考虑自动注入特定前缀的 env 环境配置

例如自动注入 `UMI_ENV_` 前缀的 env 配置，用户可以通过 `cross-env` 或者 `.env` 文件配置

可参考的框架按前缀自动注入行为：
- CRA 是 [REACT_APP_](https://create-react-app.dev/docs/adding-custom-environment-variables/)
- nextjs 是 [NEXT_PUBLIC_](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser)
- nuxtjs 是 [NUXT_ENV_](https://nuxtjs.org/docs/configuration-glossary/configuration-env#automatic-injection-of-environment-variables)
- vite 是 [VITE_](https://cn.vitejs.dev/guide/env-and-mode.html#env-files)
- vue-cli 是 [VUE_APP_](https://cli.vuejs.org/guide/mode-and-env.html#environment-variables)

想法来源：https://github.com/umijs/umi/issues/7799

**Reply by fz6m** [reply]
1. 目前 `UMI_ENV` 用于区别加载的配置文件、环境变量文件。

2. 现在 `UMI_APP_` 开头的环境变量是会被自动加载到 define 里的。

**Reply by CJY0208** [reply]
了解了，在源码中找到了 `UMI_APP_` 自动注入的部分，此前在文档里似乎没找到相关信息
